core/trino-main/src/main/java/io/trino/operator/join/unspilled/JoinProbe.java

```diff
@@ -52,22 +55,24 @@ public JoinProbe createJoinProbe(Page page, LookupSource lookupSource)
     private final int[] probeOutputChannels;
     private final int positionCount;
     private final Page page;
-    private final Page probePage;
-    @Nullable
-    private final Block probeHashBlock;
-    private final boolean probeMayHaveNull;
-    private final LookupSource lookupSource;
+    private final List<Block> nullableBlocks;
+    private final long[] joinPositionCache;
     private int position = -1;
 
     private JoinProbe(int[] probeOutputChannels, Page page, Page probePage, LookupSource lookupSource, @Nullable Block probeHashBlock)
     {
         this.probeOutputChannels = requireNonNull(probeOutputChannels, "probeOutputChannels is null");
         this.page = requireNonNull(page, "page is null");
         this.positionCount = page.getPositionCount();
-        this.probePage = requireNonNull(probePage, "probePage is null");
-        this.lookupSource = requireNonNull(lookupSource, "lookupSource is null");
-        this.probeHashBlock = probeHashBlock;
-        this.probeMayHaveNull = probeMayHaveNull(probePage);
+        requireNonNull(probePage, "probePage is null");
+        requireNonNull(lookupSource, "lookupSource is null");
+
+        nullableBlocks = (List<Block>) IntStream.range(0, probePage.getChannelCount())
```

Comment:
does `nullableBlocks` need to be a field?

Reply:
also, it seems strange to pass both `probePage` and calculated from this `nullableBlocks` as `fillCache` args.
Consider moving this logic to `fillCache`